lib/mdw_fused_dslash.cu

```diff
@@ -0,0 +1,757 @@
+      const int parity;       // output parity of this dslash operator
+      const int volume_cb;    // checkerboarded volume
+      const int volume_4d_cb; // 4-d checkerboarded volume
+
+      const int dim[4];
+      const int shift[4];      // sites where we actually calculate.
+      const int halo_shift[4]; // halo means zero. When we are expanding we have halo of cs-field where values are zero.
+
+      const int_fastdiv shrinked_dim[4]; // dimension after shifts are considered.
+
+      // partial kernel and expansion parameters
+      const int volume_4d_cb_shift; // number of 4d sites we need calculate
+      // const int volume_4d_cb_expansive; //
+
+      const real m_f; // fermion mass parameter
+      const real m_5; // Wilson mass shift
+
+      const bool dagger; // dagger
+      //    const bool xpay;        // whether we are doing xpay or not
+
+      real b; // real constant Mobius coefficient
+      real c; // real constant Mobius coefficient
+      real a; // real xpay coefficient
+
+      real kappa;
+      real fac_inv;
+
+      // (beta + alpha*m5inv) * in
+      real alpha = 1.;
+      real beta = 0.;
+
+      real m_scale = 1.; // scale factor for the matrix
+
+      bool small_kappa = false;
+
+      MdwfFusedDslashType type;
+      FusedDslashArg(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &U, ColorSpinorField &y,
+                     const ColorSpinorField &x, double m_f_, double m_5_, const Complex *b_5, const Complex *c_5,
+                     bool dagger_, int parity, int shift_[4], int halo_shift_[4], MdwfFusedDslashType type_) :
+        out(out),
+        in(in),
+        U(U),
+        y(y),
+        x(x),
+        nParity(in.SiteSubset()),
+        parity(parity),
+        volume_cb(in.VolumeCB() > out.VolumeCB() ? in.VolumeCB() : out.VolumeCB()),
+        volume_4d_cb(volume_cb / Ls_),
+        m_f(m_f_),
+        m_5(m_5_),
+        dagger(dagger_),
+        shift{shift_[0], shift_[1], shift_[2], shift_[3]},
+        halo_shift{halo_shift_[0], halo_shift_[1], halo_shift_[2], halo_shift_[3]},
+        dim{(3 - nParity) * (in.VolumeCB() > out.VolumeCB() ? in.X(0) : out.X(0)),
+            in.VolumeCB() > out.VolumeCB() ? in.X(1) : out.X(1), in.VolumeCB() > out.VolumeCB() ? in.X(2) : out.X(2),
+            in.VolumeCB() > out.VolumeCB() ? in.X(3) : out.X(3)},
+        shrinked_dim{dim[0] - 2 * shift[0], dim[1] - 2 * shift[1], dim[2] - 2 * shift[2], dim[3] - 2 * shift[3]},
+        volume_4d_cb_shift(shrinked_dim[0] * shrinked_dim[1] * shrinked_dim[2] * shrinked_dim[3] / 2),
+        type(type_)
+      {
+        if (in.Nspin() != 4) { errorQuda("nSpin = %d NOT supported.\n", in.Nspin()); }
+
+        if (nParity == 2) { errorQuda("nParity = 2 NOT supported, yet.\n"); }
+
+        // if (!in.isNative() || !out.isNative())
+        //   errorQuda("Unsupported field order out=%d in=%d\n", out.FieldOrder(), in.FieldOrder());
+
+        b = b_5[0].real();
```

Comment:
Add a check to make sure these are real, constant (again, I'm content with just comparing the first and second elements)